### ISSUE TITLE
clarify metadata patching for old python-dependent packages

### ISSUE THREAD

**Issue by mbargull** [msarahan/pypy-python-structure#3]
It's not clear to me what mechanisms will be at play to prevent old `python`-dependent packages from picking up a `python=*=pypy`. The `track_features` alone is not enough to ensure that.

**Comment by msarahan** [comment]
hotfix that adds _python_impl metapackage as dep for all legacy (non-metapackage) `python` packages

**Comment by mbargull** [comment]
We have to be careful here. If we just add a dependency, then `conda` will complain to the user that the environment is f... inconsistent due to the `python` package.
I don't really know how `conda` (current and older versions that the user may have installed) handles "inconsistent packages", TBH.

**Comment by msarahan** [comment]
Good point.  We'll need to test that carefully before rolling out.  Current
conda is a tiny bit smarter than it used to be in the case of
inconsistency, but it's far from bulletproof.

**Comment by mbargull** [comment]
And even if `python` has that dependency added, this would not prevent an existing CPython-specific package with
```json
{
"depends": ["python >=3.6,<3.7"]
}
```
from picking up `python=3.6=pypy`. Yes, `track_features` would make this less likely, but wouldn't prevent it. (E.g., if `python=3.6=pypy` is compatible with a newer `zlib` and older `python` builds aren't, but that newer `zlib` is required for the environment, `python=3.6=pypy` would be the only viable choice, hence `track_features` get ignored.)

**Comment by msarahan** [comment]
You're definitely correct on that analysis, but I think it's unlikely to cause problems.  At least not likely enough to motivate me into figuring out anything better.

**Comment by mbargull** [comment]
The simplest solution is to make `python=*=pypy` uninstallable by default by adding a dependency not found in the regular channels, e.g., `conda-forge/label/pypy::whatever`.
That may not fly well with the people wanting `pypy`, though.
It would be my preferred solution, since it is logically sound wrt. solver stuff. But that'd need much more work to make it user friendly (1. that concept is uncommon for `defaults` and `conda-forge`, 2. not supported by conda-forge build system, 3. should have support in `conda` for notifications à la "you need to add channel X for that package", ... oh well)

=> I haven't come up with a better solution that is readily deployable yet either.

**Comment by mbargull** [comment]
(I've forgotten most of a previous discussion by now and am not sure if it's much relevant here; putting it here to remind myself:)
Tangentially related: https://github.com/conda/conda/issues/6950#issuecomment-369278274

**Comment by h-vetinari** [comment]
How about a virtual package, like conda/conda#9358, conda/conda#9349 conda/conda#8267?

That would have to be implemented in conda, but then you'd at least have more control in the sense that "to install `python=*=pypy`, you need `conda >= 4.x`".

And the pypy-package could depend on that virtual package, which should take care of the solver issues.

The only point would be that the package is not really all that virtual, and should be easily changeable by the user. But just as a thought.

**Comment by mbargull** [comment]
@h-vetinari, adding a virtual package for the `conda` version itself is a magnificent idea, generally (for ensuring feature compatibilities between packages and the package manager itself)! I hadn't thought of something like that before; simple, yet elegant -- that makes me smile, thanks!
Can you open an issue on https://github.com/conda/conda/issues for this and tag me in it?

**Comment by mbargull** [comment]
> which should take care of the solver issues.

I don't think it will, though.
Acc. to
- @msarahan in https://github.com/msarahan/pypy-python-structure/issues/3#issuecomment-556199798:
> Current conda is a tiny bit smarter than it used to be in the case of inconsistency, but it's far from bulletproof.

- @jjhelmus in https://github.com/conda-forge/conda-forge.github.io/issues/867#issuecomment-556482140
> become inconsistent and the solver would be free to install just about anything.

=> even current versions of `conda` will have problems. Plus I'm not sure if it is generally possible to handle these cases in a somewhat logically consistent way, generally (without adding ugly `python`-specific special handling -- which I would oppose due to maintainability and consistency concerns).

**Comment by h-vetinari** [comment]
> > @h-vetinari:  How about a virtual package, like conda/conda#9358, conda/conda#9349 conda/conda#8267?
> > 
> > That would have to be implemented in conda, but then you'd at least have more control in the sense that "to install `python=*=pypy`, you need `conda >= 4.x`".
> 
> @mbargull: @h-vetinari, adding a virtual package for the `conda` version itself is a magnificent idea, generally (for ensuring feature compatibilities between packages and the package manager itself)! I hadn't thought of something like that before; simple, yet elegant -- that makes me smile, thanks!
Can you open an issue on https://github.com/conda/conda/issues for this and tag me in it?

Ah, I guess I was too vague (assuming you didn't mistype). I had been thinking about a virtual package for the python-implementation (and since that would have to be implemented and release in conda, it would automatically give a restriction on the conda version). And if the conda-forge `pypy` package were to depend on that, then it shouldn't need metadata hotfixes or mess with the solver IIUC - i.e. older conda versions would look for the virtual package but wouldn't find it, forcing them to stay on cpython.

In any case, you seem to have read my ambiguously worded suggestion as a virtual package for the conda version? I'm not sure I follow the idea well enough to open the issue like you asked (I mean, conda already has a version...).
